File: napkin/src/patchpanel/operatoritem.cpp
#include "operatoritem.h"
#include <nap/object.h>
#include <QGraphicsScene>
#include <QPainter>
#include "../napkin_utilities.h"
#include <QStyleOption>

inline void moveToFront(QGraphicsItem* item)
{
	QList<QGraphicsItem*> items = item->scene()->items();
	qreal highest = -10000000;
	foreach (QGraphicsItem* item, items)
		highest = qMax(highest, item->zValue());
	item->setZValue(highest + 1);
}

OperatorItem::OperatorItem(QGraphicsItem* parent, nap::Operator& op) : QGraphicsObject(parent), mOperator(op)
{
	setPos(getObjectEditorPosition(op));

	mTitleLabel = new SmallTextItem(this);
	setFlag(QGraphicsItem::ItemIsFocusable, true);
	setFlag(QGraphicsItem::ItemIsSelectable, true);
	// setFlag(QGraphicsItem::ItemIsMovable, true);

	operatorBorder = 1;
	operatorBorderSelected = 2;
//
//	QGraphicsDropShadowEffect* shadow = new QGraphicsDropShadowEffect(this);
//	shadow->setBlurRadius(8.0);
//	shadow->setColor(QColor(0x00, 0x10, 0x20, 0x63));
//	shadow->setOffset(1, 2);
//	shadow->setEnabled(true);
//	setGraphicsEffect(shadow);

	setTitle(QString::fromStdString(op.getName()));

    // TODO: THESE SIGNALS HAVE BEEN REMOVED? (COEN)
    /*
	mOperator.childAdded.connect(onInputPlugAddedSlot);
	mOperator.childAdded.connect(onOutputPlugAddedSlot);
	mOperator.childRemoved.connect(onPlugRemovedSlot);
	mOperator.childRemoved.connect(onPlugRemovedSlot);
    */

    // Listen for stuff like position changes
    // TODO: This must be made simpler
    {
        mOperator.childAdded.connect([&](nap::Object& ob) {
            if (!ob.getTypeInfo().isKindOf<nap::AttributeBase>()) {
                nap::AttributeBase& attrib = static_cast<nap::AttributeBase&>(ob);
                attrib.valueChanged.connect([=](nap::AttributeBase& at) {
                    setPos(getObjectEditorPosition(mOperator));
                });
            }
        });

        for (nap::AttributeBase* attrib : mOperator.getAttributes()) {
            attrib->valueChanged.connect([&](nap::AttributeBase& at) {
                setPos(getObjectEditorPosition(mOperator));
            });
        }
    }

	for (auto plug : mOperator.getInputPlugs())
		onInputPlugAdded(*plug);
	for (auto plug : mOperator.getOutputPlugs())
		onOutputPlugAdded(*plug);

	layout();
}

OperatorItem::~OperatorItem()
{
	/*
	mOperator->plugAdded.Disconnect(this, &OperatorItem::onPlugRemoved);
	mOperator->plugRemoved.Disconnect(this, &OperatorItem::onPlugRemoved);
	mOperator->changed.Disconnect(this, &OperatorItem::onAttributeChanged);
	*/
}


QRectF OperatorItem::boundingRect() const
{
	qreal b = isSelected() ? operatorBorderSelected : operatorBorder;
	b = operatorBorder;
	return childrenBoundingRect().adjusted(-b, -b, b, b);
}


void OperatorItem::paint(QPainter* painter, const QStyleOptionGraphicsItem* option, QWidget* widget)
{
	QStyleOptionGraphicsItem op = *option;

//	painter->drawRect(boundingRect());

	//	if (isSelected()) {
//	painter->setBrush(isSelected() ? QColor(0xFF, 0x88, 0x00) : QColor(0x80, 0x80, 0x80));
	//	}

//	painter->setBrush(NetworkStyle::Col_Node);
//	painter->drawRect(childrenBoundingRect());

	painter->setBrush(NetworkStyle::Col_Node_Title);
	painter->drawRect(mTitleRect);

	painter->setBrush(QApplication::palette().window());
	if (isSelected()) {
		painter->setPen(QPen(QApplication::palette().highlight(), 0));
	}else {
		painter->setPen(QPen(QApplication::palette().dark(), 0));
	}
	painter->drawRect(boundingRect());





//	QApplication::style()->drawPrimitive(QStyle::PE_FrameDefaultButton, option, painter, widget);
}

void OperatorItem::setTitle(const QString name)
{
	mTitleLabel->setPlainText(name);
	layout();
}


void OperatorItem::hideIncompatiblePlugs(PlugItem* src)
{
	for (PlugItem* plug : mInputPlugs) {
		if (plug == src) continue; // Do show the originating plug
		if (!canConnect(plug->plug(), src->plug())) plug->setUsable(false);
	}
	for (PlugItem* plug : mOutputPlugs) {
		if (plug == src) continue; // Do show the originating plug
		if (!canConnect(plug->plug(), src->plug())) plug->setUsable(false);
	}
}

void OperatorItem::showAllPlugs()
{
	for (PlugItem* plug : mInputPlugs)
		plug->setUsable(true);
	for (PlugItem* plug : mOutputPlugs)
		plug->setUsable(true);
}

void OperatorItem::layout()
{
	const qreal headerHeight = 20;
	const qreal headerSpacing = 6;
	const qreal itemSpacing = 14;

	qreal currentX = 0;

	// Inputs
	QList<PlugItem*> inPlugs = mInputPlugs;
	for (int i = 0, len = inPlugs.length(); i < len; i++) {
		PlugItem* item = inPlugs[i];
		item->setPos(0, headerHeight + headerSpacing + i * itemSpacing);
		currentX = qMax(currentX, item->boundingRect().width());
	}

	qreal maxX = currentX;
	// Outputs
	QList<PlugItem*> outPlugs = mOutputPlugs;
	for (int i = 0, len = outPlugs.length(); i < len; i++) {
		maxX = qMax(maxX, currentX + outPlugs[i]->boundingRect().width());
	}


	mTitleRect = mTitleLabel->boundingRect();
	mTitleRect.setWidth(childrenBoundingRect().width());
	mTitleLabel->setPos(mTitleRect.right() - mTitleLabel->boundingRect().width(), 0);

	maxX = qMax(maxX, mTitleRect.width());

	for (int i = 0, len = outPlugs.length(); i < len; i++) {
		PlugItem* item = outPlugs[i];
		float w = item->boundingRect().width();
		item->setPos(maxX - w, headerHeight + headerSpacing + i * itemSpacing);
	}
}

QVariant OperatorItem::itemChange(GraphicsItemChange change, const QVariant& value)
{
	if (change == QGraphicsItem::ItemSelectedHasChanged) moveToFront(this);

	return QGraphicsItem::itemChange(change, value);
}

void OperatorItem::onInputPlugAdded(nap::InputPlugBase& plug)
{
	// Listen for plug connections
	plug.connectedSignal.connect([&](nap::OutputPlugBase& outPlug) {
		plugConnected(outPlug, plug);
	});
	plug.disconnectedSignal.connect([&](nap::OutputPlugBase& outPlug) {
		plugDisconnected(outPlug, plug);
	});

	auto plugItem = new PlugItem(this, plug);
	mInputPlugs << plugItem;
	layout();
}

void OperatorItem::onOutputPlugAdded(nap::OutputPlugBase& plug)
{
	auto plugItem = new PlugItem(this, plug);
	mOutputPlugs << plugItem;
	layout();
}

void OperatorItem::onPlugRemoved(nap::Plug& plug) { delete findPlugItem(plug); }

void OperatorItem::onAttributeChanged(nap::AttributeBase& attrib) { setPos(getObjectEditorPosition(mOperator)); }

PlugItem* OperatorItem::findPlugItem(const nap::Plug& plug) const
{
	for (auto item : mInputPlugs)
		if (&item->plug() == &plug) return item;
	for (auto item : mOutputPlugs)
		if (&item->plug() == &plug) return item;
	return nullptr;
}

const std::string& OperatorItem::name() const { return mOperator.getName(); }


QRectF SmallTextItem::boundingRect() const { return QGraphicsTextItem::boundingRect().adjusted(0, 0, -2, -2); }


ValueItem::ValueItem(PlugItem* plug) : QGraphicsTextItem(), mPlug(plug)
{
	/*
	InPlug* inPlug = qobject_cast<InPlug*>(plug->plug());
	//Q_ASSERT(inPlug);
	setPlainText(inPlug->value().toString());
	*/


}

File: py/patch/operatoritem.py
from PyQt5.QtCore import pyqtSignal, QRectF, Qt
from PyQt5.QtGui import QPen
from PyQt5.QtWidgets import QGraphicsObject, QGraphicsTextItem, QGraphicsItem, QApplication

import nap
from patch.plugitem import PlugItem
from patch.patchutils import COL_NODE_TITLE, _canConnect, moveToFront, _getObjectEditorPos


class OperatorItem(QGraphicsObject):
    moved = pyqtSignal()
    plugConnected = pyqtSignal(object, object)
    plugDisconnected = pyqtSignal(object, object)

    def __init__(self, parent, op):
        super(OperatorItem, self).__init__()
        assert (isinstance(op, nap.Operator))
        self.__inputPlugs = []
        self.__outputPlugs = []
        self.__operator = op
        self.__titleLabel = QGraphicsTextItem(self)
        self.setFlag(QGraphicsItem.ItemIsFocusable, True)
        self.setFlag(QGraphicsItem.ItemIsSelectable, True)

        self.__operatorBorder = 1
        self.__operatorBorderSelected = 2

        self.setTitle(op.name())

        # TODO: Port some extra code from the C++ implementation here

        for p in self.__operator.inputPlugs():
            self.__onInputPlugAdded(p)
        for p in self.__operator.outputPlugs():
            self.__onOutputPlugAdded(p)

    def inputPlugs(self):
        return self.__inputPlugs

    def outputPlugs(self):
        return self.__outputPlugs

    def plugs(self):
        return self.__inputPlugs + self.__outputPlugs

    def boundingRect(self):
        b = self.__operatorBorder
        return self.childrenBoundingRect().adjusted(-b, -b, b, b)

    def paint(self, painter, option, widget=None):
        painter.setBrush(COL_NODE_TITLE)
        painter.drawRect(self.__titleRect)
        painter.setBrush(QApplication.palette().window())
        if self.isSelected():
            painter.setPen(QPen(QApplication.palette().highlight(), 0))
        else:
            painter.setPen(QPen(QApplication.palette().dark(), 0))
        painter.drawRect(self.boundingRect())

        headerRect = QRectF(0, 0, self.boundingRect().width(),
                            self.__titleRect.height())
        painter.setBrush(QApplication.palette().window().color().darker(110))
        painter.setPen(QPen(Qt.NoPen))
        painter.drawRect(headerRect.adjusted(0, 0, -2, -1))

    def setTitle(self, name):
        self.__titleLabel.setPlainText(name)
        self.layout()

    def findPlugItem(self, plug):
        for item in self.plugs():
            if item.plug() == plug:
                return item

    def name(self):
        return str(self.__titleLabel.toPlainText())

    def operator(self):
        """
        @rtype: nap.Operator
        """
        return self.__operator

    def hideIncompatiblePlugs(self, src):
        for plug in self.plugs():
            if plug == src:
                continue
            if not _canConnect(plug.plug(), src.plug()):
                plug.setUsable(False)

    def showAllPlugs(self):
        for plug in self.plugs():
            plug.setUsable(True)

    def layout(self):
        headerHeight = 20
        headerSpacing = 6
        itemSpacing = 14
        currentX = 0

        for i in range(len(self.__inputPlugs)):
            item = self.__inputPlugs[i]
            item.setPos(0, headerHeight + headerSpacing + i * itemSpacing)
            currentX = max(currentX, item.boundingRect().width())
        maxX = currentX
        for i in range(len(self.__outputPlugs)):
            maxX = max(maxX,
                       currentX + self.__outputPlugs[i].boundingRect().width())

        self.__titleRect = self.__titleLabel.boundingRect()
        self.__titleRect.setWidth(self.childrenBoundingRect().width())
        self.__titleLabel.setPos(
            self.__titleRect.right() - self.__titleLabel.boundingRect().width(),
            0)

        maxX = max(maxX, self.__titleRect.width())

        for i in range(len(self.__outputPlugs)):
            item = self.__outputPlugs[i]
            item.setPos(maxX - item.boundingRect().width(),
                        headerHeight + headerSpacing + i * itemSpacing)

    def itemChange(self, change, value):
        if change == QGraphicsItem.ItemSelectedHasChanged:
            moveToFront(self)
        if change == QGraphicsItem.ItemPositionHasChanged:
            self.moved.emit()
        return super(OperatorItem, self).itemChange(change, value)

    def __onInputPlugAdded(self, plug):
        """
        @type plug: nap.InputPlugBase
        """
        plug.connected.connect(self.__onPlugConnected)
        plug.disconnected.connect(self.__onPlugDisconnected)
        plugItem = PlugItem(self, plug)
        self.__inputPlugs.append(plugItem)
        self.layout()

    def __onOutputPlugAdded(self, plug):
        plugItem = PlugItem(self, plug)
        self.__outputPlugs.append(plugItem)
        self.layout()

    def __onPlugRemoved(self, plug):
        item = self.findPlugItem(plug)
        del item

    def __onAttributeChanged(self, attrib):
        # TODO: Review
        self.setPos(_getObjectEditorPos(self.__operator))

    def __onPlugConnected(self, srcPlug, dstPlug):
        self.plugConnected.emit(srcPlug, dstPlug)

    def __onPlugDisconnected(self, srcPlug, dstPlug):
        self.plugDisconnected.emit(srcPlug, dstPlug)